File: solutions/0746. Min Cost Climbing Stairs/0746_2.cpp
// Time complexity: O(n)
// Space complexity: O(1)
class Solution {
 public:
  int minCostClimbingStairs(const vector<int>& cost) {
    int prev = cost[0];
    int curr = cost[1];

    for (int i = 2; i < cost.size(); ++i) {
      int next = cost[i] + min(prev, curr);
      prev = curr;
      curr = next;
    }

    return min(prev, curr);
  }
};

File: solutions/0746. Min Cost Climbing Stairs/0746_2.py
# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
  def minCostClimbingStairs(self, cost: List[int]) -> int:
    prev = cost[0]
    curr = cost[1]

    for i in range(2,len(cost)):
      next = cost[i] + min(prev, curr)
      prev, curr = curr, next

    return min(prev, curr)